interface/connection/conn.go

```diff
@@ -4,55 +4,97 @@
 package connection
 
 import (
+	"fmt"
 	"io"
 	"net"
+	"os"
 	"sync"
 )
 
+const defaultUnixSocketAddress = "/run/parsec/parsec.sock"
+const parsecEndpointEnvironmentVariable = "PARSEC_SERVICE_ENDPOINT"
+
+// Connection represents a connection to the parsec service.
 type Connection interface {
+	// Open should be called before use
 	Open() error
+	// Methods to read, write and close the connection
 	io.ReadWriteCloser
 }
 
+// type to manage unix socket connection
 type unixConnection struct {
 	sync.Mutex
-	rwc io.ReadWriteCloser
-	// path string
+	rwc     io.ReadWriteCloser
+	address string
+	isOpen  bool
 }
 
+// Read data from
 func (conn *unixConnection) Read(p []byte) (n int, err error) {
+	conn.Lock()
+	defer conn.Unlock()
+	if !conn.isOpen {
+		return 0, fmt.Errorf("read called on closed connection")
+	}
 	return conn.rwc.Read(p)
 }
 
+// Write data to unix socket
 func (conn *unixConnection) Write(p []byte) (n int, err error) {
+	conn.Lock()
+	defer conn.Unlock()
+	if !conn.isOpen {
+		return 0, fmt.Errorf("write called on closed connection")
+	}
 	return conn.rwc.Write(p)
 }
 
+// Close the unix socket
 func (conn *unixConnection) Close() error {
 	conn.Lock()
 	defer conn.Unlock()
+	// We'll allow closing a closed connection
 	if conn.rwc != nil {
 		err := conn.rwc.Close()
 		if err != nil {
 			return err
 		}
 	}
 	conn.rwc = nil
+	conn.isOpen = false
 	return nil
 }
 
+// Opens the unix socket ready for read/write
+// Will attempt to look up socket from environment in PARSEC_SERVICE_ENDPOINT variable,
+// but if cannot be found, will default to /run/parsec/parsec.sock
```

Comment:
Maybe this comment is better placed on the `NewDefaultConnection` function?

Reply:
done